sgt-back-end/server.ts

```diff
@@ -0,0 +1,209 @@
+app.use(express.json());
+
+// Getting the id of the individual grade
+app.get('/api/grades/:gradeId', async (req, res, next) => {
+  try {
+    // validate the inputs FIRST
+    const gradeId = Number(req.params.gradeId);
+    if (!Number.isInteger(gradeId) || gradeId <= 0) {
+      throw new ClientError(400, '"gradeId" must be a positive integer');
+    }
+    // Ok, the input is reasonable, time to query the database.
+    // We are NOT putting the user input directly into our query;
+    // instead, we are sending the user input in a separate array.
+    /*
+     * Review the documentation on parameterized queries here:
+     * https://node-postgres.com/features/queries#parameterized-query
+     * You'll be using this technique to prevent SQL injection attacks.
+     *
+     * https://www.youtube.com/watch?v=_jKylhJtPmI
+     */
+    const sql = `
+      select *
+        from "grades"
+        where "gradeId" = $1
+    `;
+    const params = [gradeId];
+    const result = await db.query(sql, params);
+    // If we get here, the query succeeded, even if nothing was found.
+    // The Result object will include an array of rows.
+    // See the docs on results: https://node-postgres.com/apis/result
+    const grade = result.rows[0];
+    if (!grade) {
+      // We could not have known ahead of time without actually querying the db,
+      // but the specific grade being requested was not found in the database.
+      throw new ClientError(404, `Cannot find grade with "gradeId" ${gradeId}`);
+    }
+    // The specific grade was found in the database, yay! Return it.
+    res.json(grade);
+  } catch (err) {
+    // The query failed for some reason
+    // possibly due to a syntax error in the SQL statement;
+    // send `err` to the error handling middleware
+    next(err);
+  }
+});
+
+// GET REQUEST calling the grades table(SQL)
+app.get('/api/grades', async (_, res, next) => {
+  try {
+    // SQL query to get all rows from the grades table
+    const sql = 'SELECT * FROM grades';
+    // const gradesInfo = sql
+
+    // query by using the pool clause
+
+    const queryState = await db.query(sql);
+
+    const queryResult = queryState.rows[0];
+
+    res.json(queryResult);
+  } catch (err) {
+    // The query failed for some reason
+    // possibly due to a syntax error in the SQL statement;
+    // send `err` to the error handling middleware
+    next(err);
+  }
+});
+
+// Creating a new post using the POST method
+
+app.post('/api/grades', async (req, res, next) => {
+  try {
+    // SQL query to get all rows from the grades table
+    const { name, course, score } = req.body;
+
+    if (!name) {
+      throw new ClientError(400, 'no name');
+    }
+    if (!course) {
+      throw new ClientError(400, 'no course');
+    }
+    if (!score) {
+      throw new ClientError(400, 'no score');
+    }
+
+    const sql = `
+    INSERT INTO "grades" ("name" , "course" , "score" )
+      VALUES ($1 ,$2 ,$3)
+        RETURNING $1 , $2 , $3
```

Comment:
This doesn't work. Did you try it? If you don't run your code before submitting it, you will be sending me buggy code. Even worse, if you do that in a real project, you'll be making your customers run buggy code -- and they won't stay customers for very long!

The `returning` statement needs to be followed by column names. `$1`, `$2`, etc. are values, not column names. You should almost always use `returning *`. There is only one security situation you will encounter in class where you do not want to use `returning *` and that is when it would return a hashed password.